services/motion/builtin/builtin.go

```diff
@@ -237,71 +247,191 @@ func (ms *builtIn) MoveOnGlobe(
+	defer close(replanChan)
+	replanChan <- true
+
+	// errChan is for sending error messages encountered during the move
+	errChan := make(chan error)
+	defer close(errChan)
+
+	// create a waitgroup to manage goroutines and wait on their completion before exiting the function
+	var backgroundWorkers sync.WaitGroup
+	defer backgroundWorkers.Wait()
+
+	// create a cancellable context and ensure that it is cancelled before exiting
+	var cancelCtx context.Context
+	var cancelFn context.CancelFunc
+	defer func() {
+		if cancelFn != nil {
+			cancelFn()
+		}
+	}()
+
+	// helper function to manage polling functions
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
+		backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			ticker := time.NewTicker(period)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					if err := fn(ctx); err != nil {
+						errChan <- err
+					}
+					return
 				}
+			}
+		}, backgroundWorkers.Done)
+	}
+
+	// start the loop that (re)plans when something is read from the replan channel
+	// and exits when something is read from the success channel
+	for {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case <-successChan:
+			return true, nil
+		case err := <-errChan:
+			return false, err
+		case <-replanChan:
+			// cancel the goroutines spawned by this function, create a new cancellable context to use in the future
+			if cancelFn != nil {
+				cancelFn()
+			}
+			backgroundWorkers.Wait()
+			// context is not lost because it is cancelled above and then cancelled again with defer
+			//nolint:govet
+			cancelCtx, cancelFn = context.WithCancel(ctx)
+
+			inputs, err := kb.CurrentInputs(ctx)
+			if err != nil {
+				//nolint:govet
+				return false, err
+			}
+			// TODO: this is really hacky and we should figure out a better place to store this information
+			if len(kb.Kinematics().DoF()) == 2 {
+				inputs = inputs[:2]
+			}
+			planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName.Name: inputs}
+
+			plan, err := motionplan.PlanMotion(ctx, planRequest)
+			if err != nil {
 				return false, err
 			}
+
+			// drain channels of any extra messages
+			for len(replanChan) > 0 {
+				<-replanChan
+			}
+			for len(errChan) > 0 {
+				<-errChan
+			}
+
+			// spawn two goroutines that each have the ability to trigger a replan
+			// TODO: optionally could make MotionConfiguration more robust such that the option that these are zero is never possible
+			if positionPollingPeriod > 0 {
+				startPolling(cancelCtx, positionPollingPeriod, func(ctx context.Context) error {
+					// TODO: the function that actually monitors position
+					return nil
+				})
+			}
+			if obstaclePollingPeriod > 0 {
+				startPolling(cancelCtx, obstaclePollingPeriod, func(ctx context.Context) error {
+					// TODO: the function that actually monitors obstacles
+					replanChan <- true
```

Comment:
[q] does this need to remain here? could be confusing for other future work.. is there here for tests?

Reply:
Good question.. I did this for tests but when we implement the function (which will be very soon )this will ned to change